packages/gestalt/src/Button.js

```diff
@@ -28,54 +29,65 @@ const DEFAULT_TEXT_COLORS = {
   white: 'darkGray',
 };
 
-const SIZE_NAME_TO_PIXEL = {
-  sm: 10,
-  md: 12,
-  lg: 12,
-};
-
-type Props = {|
-  accessibilityControls?: string,
-  accessibilityExpanded?: boolean,
-  accessibilityHaspopup?: boolean,
+type BaseButton = {|
   accessibilityLabel?: string,
   color?: 'gray' | 'red' | 'blue' | 'transparent' | 'white',
   disabled?: boolean,
-  iconEnd?: $Keys<typeof icons>,
   inline?: boolean,
-  name?: string,
-  onClick?: AbstractEventHandler<SyntheticMouseEvent<HTMLButtonElement>>,
-  selected?: boolean,
+  onClick?: AbstractEventHandler<
+    | SyntheticMouseEvent<HTMLButtonElement>
+    | SyntheticMouseEvent<HTMLAnchorElement>
+  >,
   size?: 'sm' | 'md' | 'lg',
   text: string,
   textColor?: 'white' | 'darkGray' | 'blue' | 'red',
-  type?: 'submit' | 'button',
 |};
 
+type ButtonType = {|
+  ...BaseButton,
+  accessibilityControls?: string,
+  accessibilityExpanded?: boolean,
+  accessibilityHaspopup?: boolean,
+  iconEnd?: $Keys<typeof icons>,
+  selected?: boolean,
+  type?: 'button',
+|};
+
+type SubmitButtonType = {|
+  ...BaseButton,
+  iconEnd?: $Keys<typeof icons>,
+  name?: string,
+  type: 'submit',
+|};
+
+type LinkButtonType = {|
+  ...BaseButton,
+  href: string,
+  rel?: 'none' | 'nofollow',
+  target?: null | 'self' | 'blank',
+  type: 'link',
```

Comment:
@AlbertCarreras just so I understand. Does this mean developers would have to add `type="link"` each time they add the `href` tag?

Reply:
It means that. We could just use href to define the type we return but I think for communication clarity it's easier to explain that we have 3 types each of them requiring specific props. `submit` requires name, `link` requires href, etc. 

My thoughts, we could use just href though but href excludes type submit/button then and it has to be described rather than being explicit with type link.